include/rocksdb/utilities/object_registry.h

```diff
@@ -11,80 +11,193 @@
-// with static storage duration. For example:
-//
-//   static Registrar<Env> hdfs_reg("hdfs://.*", &CreateHdfsEnv);
-//
-// Then, calling NewCustomObject<Env>("hdfs://some_path", ...) will match the
-// regex provided above, so it returns the result of invoking CreateHdfsEnv.
-template <typename T>
-class Registrar {
+using FactoryFunc = std::function<T*(const std::string&, std::unique_ptr<T>*, std::string *)>;
+
+  
+// The ObjectRegistry is used to register objects that can be created by a
+// name/pattern at run-time where the specific implementation of the object may
+// not be known in advance.
+class ObjectRegistry {
+private:
+  // Base class for an Entry in the Registry.
+  class Entry {
+  public:
+    virtual ~Entry() {}
+    const std::string name_;  // The name of the Entry
+    Entry(const std::string& name) : name_(std::move(name)) {}
+    
+    // Checks to see if the target matches this entry
+    virtual bool matches(const std::string& target) const {
+      return name_ == target;
+    }
+  };
+  
+  // An Entry containing a FactoryFunc for creating new Objects
+  template <typename T>
+  class FactoryEntry : public Entry {
+   private:
+    std::regex pattern_;  // The pattern for this entry
+    FactoryFunc<T> factory_;
+
+   public:
+    FactoryEntry(const std::string& name, FactoryFunc<T> f)
+        : Entry(name), pattern_(std::move(name)), factory_(std::move(f)) {}
+    virtual ~FactoryEntry() {}
+    virtual bool matches(const std::string& target) const override {
+      return std::regex_match(target, pattern_);
+    }
+    // Creates a new T object.
+    T* NewFactoryObject(const std::string& target,
+			std::unique_ptr<T>* guard,
+			std::string *msg) const {
+      return factory_(target, guard, msg);
+    }
+  };
+public:
+  class Library {
+  private:
+    friend ObjectRegistry;
+    
+    // ** FactoryFunctions for this loader, organized by type
+    std::unordered_map<std::string,
+		       std::vector<std::unique_ptr<Entry> > > entries_;
+    // Finds the entry matching the input name and type
+    const Entry* FindEntry(const std::string& type,
+			   const std::string& name) const;
+    void AddEntry(const std::string & type, std::unique_ptr<Entry> & entry);
+  public:
+    void Dump(Logger *logger) const;
+    // To register a factory function for a type T, initialize a Registrar<T>
+    // object with static storage duration. For example:
+    template <typename T>
+    const FactoryFunc<T>& Register(const std::string& pattern,
+                                 const FactoryFunc<T>& factory) {
+      std::unique_ptr<Entry> entry(new FactoryEntry<T>(pattern, factory));
+      AddEntry(T::Type(), entry);
+      return factory;
+    }
+  };
+  
+private:
+  std::vector<std::shared_ptr<Library>> libraries_;
  public:
-  explicit Registrar(std::string pattern, FactoryFunc<T> factory);
-};
+  static std::shared_ptr<Library>& Default();
```

Comment:
Is `DefaultLibrary` a better name? Otherwise, I thought `ObjectRegistry::Default()` returns the default instance of `ObjectRegistry`, as `Env::Default()`.

Reply:
I moved ObjectLibrary to a higher order class and made Default a method of it.